.github/workflows/split-package-detection.yaml

```
@@ -0,0 +1,115 @@
+          declare -A seen
+          declare -A last_conflict_jar  # Track the last conflicting jar for each package
+          
+          echo "🔍 Scanning all JARs under: $ROOT_DIR (excluding test JARs)"
+          
+          # Get all JAR files first, excluding test JARs
+          jar_files=()
+          while IFS= read -r jar; do
+            # Skip JAR files ending with "-tests.jar"
+            if [[ "$jar" != *"-tests.jar" ]]; then
+              jar_files+=("$jar")
+            fi
+          done < <(find "$ROOT_DIR" -type f -name "*.jar")
+          
+          echo "Found ${#jar_files[@]} non-test JAR files to analyze"
+          
+          # Process each JAR file
+          for jar in "${jar_files[@]}"; do
+            jarname=$(realpath "$jar")
+            jarbasename=$(basename "$jarname")  # Get just the filename without path
+            tmpdir=$(mktemp -d)
+            unzip -qq "$jar" -d "$tmpdir"
+            
+            # Get all class files
+            class_files=()
+            while IFS= read -r classfile; do
+              class_files+=("$classfile")
+            done < <(find "$tmpdir" -type f -name "*.class")
+            
+            # Process each class file
+            for classfile in "${class_files[@]}"; do
+              pkg=$(dirname "${classfile#$tmpdir/}" | tr '/' '.')
+              [[ "$pkg" == "." ]] && continue  # skip default package
+            
+              if [ -n "${package_map[$pkg]}" ]; then
+                if [ "${package_map[$pkg]}" != "$jarbasename" ]; then
+                  if [[ -z "${seen[$pkg]}" ]]; then
+                    # First time seeing this conflict
+                    echo "🚨 Split package detected: $pkg"
+                    echo "    ↳ in: ${package_map[$pkg]}"
+                    echo "    ↳ and: $jarbasename"
+                    seen[$pkg]=1
+                    last_conflict_jar[$pkg]="$jarbasename"
+                  elif [[ "${last_conflict_jar[$pkg]}" != "$jarbasename" ]]; then
+                    # New jar with same conflict
+                    echo "    ↳ also in: $jarbasename"
+                    last_conflict_jar[$pkg]="$jarbasename"
+                  fi
+                  # If it's the same jar as last time, we don't print anything
+                fi
+              else
+                package_map[$pkg]=$jarbasename;
+              fi
+            done
+              
+            rm -rf "$tmpdir"
+          done
+            
+          if [[ ${#seen[@]} -eq 0 ]]; then
+            echo "✅ No split packages found!"
+          else
+            echo "❌ Split packages detected — please fix before modularizing."
+            exit 1
+          fi
+          EOF
+          chmod +x check-split-packages.sh
+
+      - name: Run split package detection
+        run: |
+          ./check-split-packages.sh 
+
+      - name: Upload results as artifact if failure
+        if: failure()
+        uses: actions/upload-artifact@v3
```

```suggestion
        uses: actions/upload-artifact@v4
```
If I recall correctly v3 is not supported any more, and it fails

---

Right, I bumped all v3 occurrences to v4.